weblab_cg/tests/test_chaste_cg_develop.py

```diff
@@ -394,6 +395,23 @@ def _check_equation_list(self, expected, generated):
                     if expected != new_expected:
                         expected = new_expected
                         del expected[expected_element_index]
+
+        # Maybe converters not recognized as converters in the expected?
```

Comment:
I don't understand what this block is doing!

Reply:
it's basically a crude way of checking whether any substitutions have happened above.

Reply:
OK, so I kind of get that it's a similar processing to the previous block, but on the newly generated code rather than the reference code from Chaste. However it's not doing exactly the same thing - should it be? Could you add examples of the kind(s) of thing it catches (in both cases actually) to illustrate why it's there?

Reply:
changed it a bit as it seems to work both ways but in the basis the issue is going from something like this from PyCml:
```
        const double var_chaste_interface__membrane__i_Stim = GetIntracellularAreaStimulus(var_chaste_interface__environment__time);
        const double var_K__I_stim_converter = var_chaste_interface__membrane__i_Stim; // uA_per_cm2
        const double var_K__chaste_interface__chaste_membrane_capacitance = HeartConfig::Instance()->GetCapacitance(); // uF_per_cm2
        const double var_K__I_stim = var_K__I_stim_converter / var_K__chaste_interface__chaste_membrane_capacitance; // uA_per_uF
```
To this which I get straight from cellmlmanip:
```
const double var_membrane__i_Stim = GetIntracellularAreaStimulus(var_chaste_interface__environment__time)/ HeartConfig::Instance()->GetCapacitance(); // uA_per_uF
```

and then in the case of `EvaluateYDerivatives`
the bit within the else there is (PyCml)
```
        if (mSetVoltageDerivativeToZero)
        {
            d_dt_chaste_interface__membrane__Vm = 0.0;
        }
        else
        {
            const double var_membrane__i_Stim_converter = var_chaste_interface__membrane__i_Stim; // uA_per_cm2
            const double var_membrane__chaste_interface__chaste_membrane_capacitance = HeartConfig::Instance()->GetCapacitance(); // uF_per_cm2
            const double var_membrane__i_Stim = var_membrane__i_Stim_converter / var_membrane__chaste_interface__chaste_membrane_capacitance; // uA_per_uF
.....
            const double var_membrane__kiont = var_membrane__IKr + var_membrane__IKs + var_membrane__IK1 + var_membrane__IKp + ((-2.0) * var_membrane__INaK) + var_membrane__Ito1 + var_membrane__i_Stim; // uA_per_uF
```
while the generated version has:
```
        if (mSetVoltageDerivativeToZero)
        {
            d_dt_chaste_interface__membrane__Vm = 0.0;
        }
        else
        {
.....
            const double var_membrane__kiont = -2.0 * var_INaK__INaK + var_IK1__IK1 + var_IKp__IKp + var_IKr__IKr + var_IKs__IKs + var_Ito1__Ito1 + var_membrane__i_Stim; // uA_per_uF
```

So that bit of code is trying to assess than that's actually fine, since var_membrane__i_Stim was already defined above and defined as the same thing.

Reply:
OK, so essentially you want to be going through extra equations on one side, checking their LHS isn't defined in some other way by the other generator (i.e. they really are extra), then substituting the definition in place of the LHS where it appears in all subsequent equations? That algorithm sounds like it should resolve the mismatch and be the same in both directions?

Reply:
And might actually mean you don't need the linker handling for the simpler cases?

Reply:
I have yet to find a real model where no linker handling is needed but in theory yes that's all accurate.
Sadly also some of the lhs are named differently, which sometimes leads to equations that are the same except for some variable names (that's attempted to resolve lower down in the check if the 2 lhs are not the same)

Reply:
Ah, yes, if the LHS names don't match up then you can't just use my straightforward algorithm to cover everything, pity.

Since this is only an interim transition stage anyway, and we have further testing by simulating the generated code, I don't think it's worth spending lots of time trying to refine the process!

Reply:
Probably not. I've find it pretty helpful though to be able to know what equation causes things not to be the same as for big models the lists can be very long so it's hard to see manually.